# Exploring Ebay Car Sales Data

## 1. Introduction:

In this project, we'll work with a dataset of used cars from `eBay Kleinanzeigen`, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to `Kaggle`. We've made a few modifications from the original dataset that was uploaded to `Kaggle`. The content of the data is in german, it's convenient to translate it first if you cannot speak german.

https://data.world/data-society/used-cars-data


The aim of this project is to clean the data and analyze the included used car listings.



### 1.2. Data dictionary:

- `dateCrawled: ` When this ad was first crawled. All field-values are taken from this date

- `name:` Name of the car

- `seller:` Whether the seller is private or a dealer

- `offerType:` The type of listing

- `price:` The price on the ad to sell the car

- `abtest:` Whether the listing is included in an A/B test

- `vehicleType:` The vehicle Type

- `yearOfRegistration:` The year in which the car was first registered

- `gearbox:` The transmission type

- `powerPS:` The power of the car in PS

- `model:` The car model name

- `kilometer:` How many kilometers the car has driven

- `monthOfRegistration:` The month in which the car was first registered

- `fuelType:` What type of fuel the car uses

- `brand:` The brand of the car

- `notRepairedDamage:` If the car has a damage which is not yet repaired

- `dateCreated:` The date on which the eBay listing was created

- `nrOfPictures:` The number of pictures in the ad

- `postalCode:` The postal code for the location of the vehicle

- `lastSeenOnline:` When the crawler saw this ad last online


In [88]:
# importing the libraries
import pandas as pd
import numpy as np

In [89]:
# reading the csv file into pandas
autos = pd.read_csv('autos.csv', encoding='Latin-1')

# rendering the variable
autos

,dateCrawled,name,seller,offerType,price,abtest,vehicleType,yearOfRegistration,gearbox,powerPS,model,odometer,monthOfRegistration,fuelType,brand,notRepairedDamage,dateCreated,nrOfPictures,postalCode,lastSeen
0,2016-03-26 17:47:46,Peugeot_807_160_NAVTECH_ON_BOARD,privat,Angebot,"$5,000",control,bus,2004,manuell,158,andere,"150,000km",3,lpg,peugeot,nein,2016-03-26 00:00:00,0,79588,2016-04-06 06:45:54
1,2016-04-04 13:38:56,BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik,privat,Angebot,"$8,500",control,limousine,1997,automatik,286,7er,"150,000km",6,benzin,bmw,nein,2016-04-04 00:00:00,0,71034,2016-04-06 14:45:08
2,2016-03-26 18:57:24,Volkswagen_Golf_1.6_United,privat,Angebot,"$8,990",test,limousine,2009,manuell,102,golf,"70,000km",7,benzin,volkswagen,nein,2016-03-26 00:00:00,0,35394,2016-04-06 20:15:37
3,2016-03-12 16:58:10,Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...,privat,Angebot,"$4,350",control,kleinwagen,2007,automatik,71,fortwo,"70,000km",6,benzin,smart,nein,2016-03-12 00:00:00,0,33729,2016-03-15 03:16:28
4,2016-04-01 14:38:50,Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfle...,privat,Angebot,"$1,350",test,kombi,2003,manuell,0,focus,"150,000km",7,benzin,ford,nein,2016-04-01 00:00:00,0,39218,2016-04-01 14:38:50
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
49995,2016-03-27 14:38:19,Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon,privat,Angebot,"$24,900",control,limousine,2011,automatik,239,q5,"100,000km",1,diesel,audi,nein,2016-03-27 00:00:00,0,82131,2016-04-01 13:47:40
49996,2016-03-28 10:50:25,Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu...,privat,Angebot,"$1,980",control,cabrio,1996,manuell,75,astra,"150,000km",5,benzin,opel,nein,2016-03-28 00:00:00,0,44807,2016-04-02 14:18:02
49997,2016-04-02 14:44:48,Fiat_500_C_1.2_Dualogic_Lounge,privat,Angebot,"$13,200",test,cabrio,2014,automatik,69,500,"5,000km",11,benzin,fiat,nein,2016-04-02 00:00:00,0,73430,2016-04-04 11:47:27
49998,2016-03-08 19:25:42,Audi_A3_2.0_TDI_Sportback_Ambition,privat,Angebot,"$22,900",control,kombi,2013,manuell,150,a3,"40,000km",11,diesel,audi,nein,2016-03-08 00:00:00,0,35683,2016-04-05 16:45:07


In [90]:
# information about the dataframe
print(autos.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

After reviewing the data, we can make the following observations:

- We see that `autos` dataset has `50,000` cars and `20` columns which most of them are strings

- Some columns have null (or NaN) values, but none have more than `~20%` null values

- The column names use `camelcase` instead of Python's preferred `snakecase`, which means that we can't just replace spaces with underscores

- The `autos` dataset is in German, it's convenient that we translate it to English in case it's necessary

- The `price` and `odometer` columns are strings instead of numeric values

- The `nrOfPictures` column only has one value, `0`

## 2. Cleaning Column Names

We will convert the column names from `camelcase` to `snakecase` and reword some of the column names based on the data dictionary to be more descriptive.

In [91]:
# printing the column names
old_columns = autos.columns
old_columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [92]:
# editing columns names
autos.rename({'dateCrawled':'date_crawled',
                    'offerType':'offer_type', 
                    'vehicleType':'vehicle_type', 
                    'yearOfRegistration':'registration_year', 
                    'powerPS':'power_ps', 
                    'monthOfRegistration':'registration_month', 
                    'fuelType':'fuel_type', 
                    'notRepairedDamage':'unrepaired_damage', 
                    'dateCreated':'ad_created', 
                    'nrOfPictures':'nr_of_pictures', 
                    'postalCode':'postal_code', 
                    'lastSeen':'last_seen'}, axis=1, inplace=True)

In [93]:
# assigning the modified columns names
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [94]:
# data set with new columns
autos.head()

,date_crawled,name,seller,offer_type,price,abtest,vehicle_type,registration_year,gearbox,power_ps,model,odometer,registration_month,fuel_type,brand,unrepaired_damage,ad_created,nr_of_pictures,postal_code,last_seen
0,2016-03-26 17:47:46,Peugeot_807_160_NAVTECH_ON_BOARD,privat,Angebot,"$5,000",control,bus,2004,manuell,158,andere,"150,000km",3,lpg,peugeot,nein,2016-03-26 00:00:00,0,79588,2016-04-06 06:45:54
1,2016-04-04 13:38:56,BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik,privat,Angebot,"$8,500",control,limousine,1997,automatik,286,7er,"150,000km",6,benzin,bmw,nein,2016-04-04 00:00:00,0,71034,2016-04-06 14:45:08
2,2016-03-26 18:57:24,Volkswagen_Golf_1.6_United,privat,Angebot,"$8,990",test,limousine,2009,manuell,102,golf,"70,000km",7,benzin,volkswagen,nein,2016-03-26 00:00:00,0,35394,2016-04-06 20:15:37
3,2016-03-12 16:58:10,Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...,privat,Angebot,"$4,350",control,kleinwagen,2007,automatik,71,fortwo,"70,000km",6,benzin,smart,nein,2016-03-12 00:00:00,0,33729,2016-03-15 03:16:28
4,2016-04-01 14:38:50,Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfle...,privat,Angebot,"$1,350",test,kombi,2003,manuell,0,focus,"150,000km",7,benzin,ford,nein,2016-04-01 00:00:00,0,39218,2016-04-01 14:38:50


Due to Python prefers working with snakecase, we converted the column names from `camelcase` to `snakecase`. Also, we edited some of the column names based on the data dictionary to be more descriptive.

## 3. Initial Exploration and Cleaning

We'll do some basic data exploration to determine what other cleaning tasks need to be done.

Initially, we will look for:

- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis

- Examples of numeric data stored as text which can be cleaned and converted

In [95]:
# Descriptive statistics for all columns
autos.describe(include='all')

,date_crawled,name,seller,offer_type,price,abtest,vehicle_type,registration_year,gearbox,power_ps,model,odometer,registration_month,fuel_type,brand,unrepaired_damage,ad_created,nr_of_pictures,postal_code,last_seen
count,50000,50000,50000,50000,50000,50000,44905,50000.000000,47320,50000.000000,47242,50000,50000.000000,45518,50000,40171,50000,50000.0,50000.000000,50000
unique,48213,38754,2,2,2357,2,8,NaN,2,NaN,245,13,NaN,7,40,2,76,NaN,NaN,39481
top,2016-04-04 16:40:33,Ford_Fiesta,privat,Angebot,$0,test,limousine,NaN,manuell,NaN,golf,"150,000km",NaN,benzin,volkswagen,nein,2016-04-03 00:00:00,NaN,NaN,2016-04-07 06:17:27
freq,3,78,49999,49999,1421,25756,12859,NaN,36993,NaN,4024,32424,NaN,30107,10687,35232,1946,NaN,NaN,8
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2005.073280,NaN,116.355920,NaN,NaN,5.723360,NaN,NaN,NaN,NaN,0.0,50813.627300,NaN
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,105.712813,NaN,209.216627,NaN,NaN,3.711984,NaN,NaN,NaN,NaN,0.0,25779.747957,NaN
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1000.000000,NaN,0.000000,NaN,NaN,0.000000,NaN,NaN,NaN,NaN,0.0,1067.000000,NaN
25%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1999.000000,NaN,70.000000,NaN,NaN,3.000000,NaN,NaN,NaN,NaN,0.0,30451.000000,NaN
50%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2003.000000,NaN,105.000000,NaN,NaN,6.000000,NaN,NaN,NaN,NaN,0.0,49577.000000,NaN
75%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2008.000000,NaN,150.000000,NaN,NaN,9.000000,NaN,NaN,NaN,NaN,0.0,71540.000000,NaN


Our initial observations:

- The columns `seller` and `offer_type` have same values which are not relevant for our analysis

- The `price` and `odometer` columns have string values. They should be converted to numeric type
    

In [96]:
# analyzing nr_of_pictures
autos['nr_of_pictures'].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

In [97]:
# analyzing seller
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [98]:
# analyzing offer_type
autos['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

As can be seen in the cell above, the column `nr_of_pictures` has only `0` values, it can be deleted from our dataset. Also, the columns `seller` and `offer_type` has only two values, but most of them are `privat` or `Angebot`; for this reason, we delete those columns from the data set.

In [99]:
# deleting columns
autos = autos.drop(['seller', 'offer_type', 'nr_of_pictures'], axis=1)

In [100]:
# identifying missing values
print(autos.isnull().sum())

date_crawled             0
name                     0
price                    0
abtest                   0
vehicle_type          5095
registration_year        0
gearbox               2680
power_ps                 0
model                 2758
odometer                 0
registration_month       0
fuel_type             4482
brand                    0
unrepaired_damage     9829
ad_created               0
postal_code              0
last_seen                0
dtype: int64


Additionally, we identified missing values. We found out that we have five columns with null values (`vehicle_type`, `gearbox`, `model`, `fuel_type`, and `unrepaired_damage`).

The `price` and `odometer` columns are numeric values stored as text.

In [101]:
# price and odometer columns
autos.loc[:5, ['price', 'odometer']]

,price,odometer
0,"$5,000","150,000km"
1,"$8,500","150,000km"
2,"$8,990","70,000km"
3,"$4,350","70,000km"
4,"$1,350","150,000km"
5,"$7,900","150,000km"


In [102]:
# removing any non-numeric character
autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '')
autos['odometer'] = autos['odometer'].str.replace(',', '').str.replace('km', '')

# converting the column to a numeric dtype
autos['price'] = autos['price'].astype(int)
autos['odometer'] = autos['odometer'].astype(int)

# renaming the column
autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)

In [103]:
# price and odometer columns
autos.loc[:5, ['price', 'odometer_km']]

,price,odometer_km
0,5000,150000
1,8500,150000
2,8990,70000
3,4350,70000
4,1350,150000
5,7900,150000


## 4. Exploring the Odometer and Price Columns

We'll continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the `odometer_km` and `price` columns.

In [104]:
# unique values
print('Unique values in "price" column: ', autos['price'].unique().shape[0])
print('\n')
print('Unique values in "odometer_km" column: ', autos['odometer_km'].unique().shape[0])

Unique values in "price" column:  2357


Unique values in "odometer_km" column:  13


In [105]:
# min/max/median/mean etc
print('Descriptive statistics - price')
print(autos['price'].describe())
print('\n')
print('Descriptive statistics - odometer_km')
print(autos['odometer_km'].describe())

Descriptive statistics - price
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


Descriptive statistics - odometer_km
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


In [106]:
# finding the counts of "price" and "odometer_km"
print('Value counts - price - descending')
print(autos['price'].value_counts().sort_index(ascending=False).head(15))
print('\n')
print('Value counts - price - ascending')
print(autos['price'].value_counts().sort_index(ascending=True).head(15))
print('\n')
print('Value counts - odometer_km')
print(autos['odometer_km'].value_counts().sort_index(ascending=False).head(10))

Value counts - price - descending
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64


Value counts - price - ascending
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
Name: price, dtype: int64


Value counts - odometer_km
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
Name: odometer_km, dtype: int64


As a result of this analizis, we'll make the following modifications in the `price` column:

- Remove `14` cars which have values above `$350,000`, because the prices increase dramatically after this value
- Remove `1,421` cars with the price of `$0`


In [107]:
# removing outlietrs
autos = autos[autos['price'].between(1, 350001)]
autos

,date_crawled,name,price,abtest,vehicle_type,registration_year,gearbox,power_ps,model,odometer_km,registration_month,fuel_type,brand,unrepaired_damage,ad_created,postal_code,last_seen
0,2016-03-26 17:47:46,Peugeot_807_160_NAVTECH_ON_BOARD,5000,control,bus,2004,manuell,158,andere,150000,3,lpg,peugeot,nein,2016-03-26 00:00:00,79588,2016-04-06 06:45:54
1,2016-04-04 13:38:56,BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik,8500,control,limousine,1997,automatik,286,7er,150000,6,benzin,bmw,nein,2016-04-04 00:00:00,71034,2016-04-06 14:45:08
2,2016-03-26 18:57:24,Volkswagen_Golf_1.6_United,8990,test,limousine,2009,manuell,102,golf,70000,7,benzin,volkswagen,nein,2016-03-26 00:00:00,35394,2016-04-06 20:15:37
3,2016-03-12 16:58:10,Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...,4350,control,kleinwagen,2007,automatik,71,fortwo,70000,6,benzin,smart,nein,2016-03-12 00:00:00,33729,2016-03-15 03:16:28
4,2016-04-01 14:38:50,Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfle...,1350,test,kombi,2003,manuell,0,focus,150000,7,benzin,ford,nein,2016-04-01 00:00:00,39218,2016-04-01 14:38:50
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
49995,2016-03-27 14:38:19,Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon,24900,control,limousine,2011,automatik,239,q5,100000,1,diesel,audi,nein,2016-03-27 00:00:00,82131,2016-04-01 13:47:40
49996,2016-03-28 10:50:25,Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu...,1980,control,cabrio,1996,manuell,75,astra,150000,5,benzin,opel,nein,2016-03-28 00:00:00,44807,2016-04-02 14:18:02
49997,2016-04-02 14:44:48,Fiat_500_C_1.2_Dualogic_Lounge,13200,test,cabrio,2014,automatik,69,500,5000,11,benzin,fiat,nein,2016-04-02 00:00:00,73430,2016-04-04 11:47:27
49998,2016-03-08 19:25:42,Audi_A3_2.0_TDI_Sportback_Ambition,22900,control,kombi,2013,manuell,150,a3,40000,11,diesel,audi,nein,2016-03-08 00:00:00,35683,2016-04-05 16:45:07


## 5. Exploring the date columns

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler

- `last_seen`: added by the crawler

- `ad_created`: from the website

- `registration_month`: from the website

- `registration_year`: from the website

Right now, `the date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. 

The other two columns are represented as numeric values, so we can use methods like `Series.describe()` to understand the distribution without any extra data processing.

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like is shown in the next cell:


In [109]:
# values in the string columns
autos[['date_crawled','ad_created','last_seen']][:5]

,date_crawled,ad_created,last_seen
0,2016-03-26 17:47:46,2016-03-26 00:00:00,2016-04-06 06:45:54
1,2016-04-04 13:38:56,2016-04-04 00:00:00,2016-04-06 14:45:08
2,2016-03-26 18:57:24,2016-03-26 00:00:00,2016-04-06 20:15:37
3,2016-03-12 16:58:10,2016-03-12 00:00:00,2016-03-15 03:16:28
4,2016-04-01 14:38:50,2016-04-01 00:00:00,2016-04-01 14:38:50


We use the workflow to calculate the distribution of values in the `date_crawled`, `ad_created`, and `last_seen` columns (all string columns) as percentages.

In [110]:
# distribution of values in 'date_crawled'
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [111]:
# distribution of values in 'ad_created'
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

In [112]:
# distribution of values in 'last_seen'
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The most of the cars values were crawled on March to April 2016 in the columns `date_crawled` and `last_seen`. On the other hand, the `ad_created` column has values that were created since June 2015 to April 2016.


In [113]:
# descriptive statistics of 'registration_year'
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The descriptive statistics analysis gives information about the minimun and the maximun year of when the cars were registerd, however it does not make sense because we have cars registered in the year 1000 and 9999.


## 6. Dealing with Incorrect Registration Year Data

One thing that stands out from the exploration we did is that the `registration_year` column contains some odd values:

- The minimum value is 1000, before cars were invented

- The maximum value is 9999, many years into the future

Because a car cannot be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

In [114]:
# deleting values outside the 1900 - 2016 interval
autos = autos[autos['registration_year'].between(1900, 2016)]

# calculating the distribution
autos['registration_year'].value_counts(normalize=True)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
          ...   
1938    0.000021
1948    0.000021
1927    0.000021
1931    0.000021
1952    0.000021
Name: registration_year, Length: 78, dtype: float64

## 7. Exploring Price by Brand

When working with data on cars it's natural to explore variations across different car brands. We can use aggregation to understand the brand column.

In [115]:
# identifying the unique values we want to aggregate by
autos['brand'].value_counts().head(20)

volkswagen        9862
bmw               5137
opel              5022
mercedes_benz     4503
audi              4041
ford              3263
renault           2201
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            654
toyota             593
hyundai            468
sonstige_autos     458
volvo              427
mini               409
Name: brand, dtype: int64

In [116]:
# assigning the index of the top 20 car brands
top_20_brands = autos['brand'].value_counts().head(20).index
top_20_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini'],
      dtype='object')

In [117]:
# creating an empty dictionary to store our aggregate data
top_20_prices = {}

# loop over the top 20 brands
for each_brand in top_20_brands:
    top_autos = autos[autos['brand'] == each_brand]
    mean_price = int(top_autos['price'].mean())                   
    top_20_prices[each_brand] = mean_price

# printing the dictionary
print(top_20_prices)

{'volkswagen': 5402, 'bmw': 8332, 'opel': 2975, 'mercedes_benz': 8628, 'audi': 9336, 'ford': 3749, 'renault': 2474, 'peugeot': 3094, 'fiat': 2813, 'seat': 4397, 'skoda': 6368, 'nissan': 4743, 'mazda': 4112, 'smart': 3580, 'citroen': 3779, 'toyota': 5167, 'hyundai': 5365, 'sonstige_autos': 12338, 'volvo': 4946, 'mini': 10613}


We observed that in the top 20 brands, there's a distinct price gap:

- Audi, BMW and Mercedes Benz are more expensive
- Renault and Fiat are less expensive
- Volkswagen is in between


## 8. Storing Aggregate Data in a DataFrame



For the top 20 brands, we use aggregation to understand the average mileage for those cars and if there's any visible link with mean price.

First, we create an empty dictionary named `top_20_mileage` to save the mean mileage per each brand.

In [118]:
# creating an empty dictionary to store our aggregate data
top_20_mileage = {}

# loop over the top 20 brands
for each_brand in top_20_brands:
    top_autos = autos[autos['brand'] == each_brand]
    mean_mileage = int(top_autos['odometer_km'].mean())
    top_20_mileage[each_brand] = mean_mileage
    
# printing the dictionary
print(top_20_mileage)

{'volkswagen': 128707, 'bmw': 132572, 'opel': 129310, 'mercedes_benz': 130788, 'audi': 129157, 'ford': 124266, 'renault': 128071, 'peugeot': 127153, 'fiat': 117121, 'seat': 121131, 'skoda': 110848, 'nissan': 118330, 'mazda': 124464, 'smart': 99326, 'citroen': 119694, 'toyota': 115944, 'hyundai': 106442, 'sonstige_autos': 89956, 'volvo': 138067, 'mini': 88105}


We'll combine the data from series objects into a single dataframe (with a shared index, `top_20_brands`) and display the dataframe directly.

In [119]:
# converting dictionaries to series objects
top_20_prices_series = pd.Series(top_20_prices)
top_20_mileage_series = pd.Series(top_20_mileage)

# creating a dataframe from the series objects
top_brands = pd.DataFrame(top_20_prices_series, columns=['mean_price'])

# assigning the series object in the dataframe
top_brands['mean_mileage'] = top_20_mileage_series

# printing the dataframe
top_brands

,mean_price,mean_mileage
volkswagen,5402,128707
bmw,8332,132572
opel,2975,129310
mercedes_benz,8628,130788
audi,9336,129157
ford,3749,124266
renault,2474,128071
peugeot,3094,127153
fiat,2813,117121
seat,4397,121131


In [120]:
# analizing the dataframe
print(top_brands.sort_values('mean_price', ascending=False))
print('\n')
print(top_brands.sort_values('mean_mileage', ascending=False))

                mean_price  mean_mileage
sonstige_autos       12338         89956
mini                 10613         88105
audi                  9336        129157
mercedes_benz         8628        130788
bmw                   8332        132572
skoda                 6368        110848
volkswagen            5402        128707
hyundai               5365        106442
toyota                5167        115944
volvo                 4946        138067
nissan                4743        118330
seat                  4397        121131
mazda                 4112        124464
citroen               3779        119694
ford                  3749        124266
smart                 3580         99326
peugeot               3094        127153
opel                  2975        129310
fiat                  2813        117121
renault               2474        128071


                mean_price  mean_mileage
volvo                 4946        138067
bmw                   8332        132572
mercedes_benz 

As can be seen in both tables, although `audi`, `mercedes_benz` and `bmw` are the most expensives cars, these brands also have the highest mean mileage.

## 9. Next Steps

In this project, we practiced applying a variety of pandas methods to explore and understand a dataset on car listings. Additionally, we'll consider the following steps:

- Data cleaning:
    - Identify categorical data that uses German words, translate them and map the values to their English counterparts
    - Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321
    - See if there are particular keywords in the name column that we can extract as new columns
    
- Analysis:
    - Find the most common brand/model combinations
    - Split the `odometer_km` into groups, and use aggregation to see if average prices follows any patterns based on the milage
    - How much cheaper are cars with damage than their non-damaged counterparts?


In [121]:
autos.head()

,date_crawled,name,price,abtest,vehicle_type,registration_year,gearbox,power_ps,model,odometer_km,registration_month,fuel_type,brand,unrepaired_damage,ad_created,postal_code,last_seen
0,2016-03-26 17:47:46,Peugeot_807_160_NAVTECH_ON_BOARD,5000,control,bus,2004,manuell,158,andere,150000,3,lpg,peugeot,nein,2016-03-26 00:00:00,79588,2016-04-06 06:45:54
1,2016-04-04 13:38:56,BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik,8500,control,limousine,1997,automatik,286,7er,150000,6,benzin,bmw,nein,2016-04-04 00:00:00,71034,2016-04-06 14:45:08
2,2016-03-26 18:57:24,Volkswagen_Golf_1.6_United,8990,test,limousine,2009,manuell,102,golf,70000,7,benzin,volkswagen,nein,2016-03-26 00:00:00,35394,2016-04-06 20:15:37
3,2016-03-12 16:58:10,Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...,4350,control,kleinwagen,2007,automatik,71,fortwo,70000,6,benzin,smart,nein,2016-03-12 00:00:00,33729,2016-03-15 03:16:28
4,2016-04-01 14:38:50,Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfle...,1350,test,kombi,2003,manuell,0,focus,150000,7,benzin,ford,nein,2016-04-01 00:00:00,39218,2016-04-01 14:38:50


### 9.1. Data cleaning:

We will do:

- Explore the data in our dataset `autos` and identify patterns and special cases    
- Remove any rows that have missing values
- Rename columns' values that have German words    
- Convert the dates columns into numeric dtype
- Explore if there are particular keywords in the `name` column that we can extract as new columns

We will identify categorical data that uses German words, translate them and map the values to their English counterparts.

In [122]:
# identifying patterns and special cases
print('Column - vehicle_type')
print(autos['vehicle_type'].unique())
print('\n')
print('Column - gearbox')
print(autos['gearbox'].unique())
print('\n')
print('Column- fuel_type')
print(autos['fuel_type'].unique())
print('\n')
print('Column - unrepaired_damage')
print(autos['unrepaired_damage'].unique())
print('\n')
print(autos.isnull().sum())

Column - vehicle_type
['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']


Column - gearbox
['manuell' 'automatik' nan]


Column- fuel_type
['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere']


Column - unrepaired_damage
['nein' nan 'ja']


date_crawled             0
name                     0
price                    0
abtest                   0
vehicle_type          2704
registration_year        0
gearbox               2110
power_ps                 0
model                 2193
odometer_km              0
registration_month       0
fuel_type             3318
brand                    0
unrepaired_damage     8307
ad_created               0
postal_code              0
last_seen                0
dtype: int64


We identified German words in four columns in our dataset (`vehicle_type`, `gearbox`, `fuel_type` and `unrepaired_damage`). In addition, these columns and `model` column contain null (or NaN) values that can be removed.

In [123]:
# removing rows with null (or NaN) values
autos = autos.dropna(axis=0)

In [124]:
# checking null values
print(autos.isnull().sum())

date_crawled          0
name                  0
price                 0
abtest                0
vehicle_type          0
registration_year     0
gearbox               0
power_ps              0
model                 0
odometer_km           0
registration_month    0
fuel_type             0
brand                 0
unrepaired_damage     0
ad_created            0
postal_code           0
last_seen             0
dtype: int64


In [125]:
# creating a dictionaries with the corrections
vehicle_type_corrections = {'bus': 'bus', 'limousine': 'limousine',
                            'kleinwagen': 'small_car', 'kombi': 'combi',
                            'coupe': 'coupe', 'suv': 'suv',
                            'cabrio': 'convertible', 'andere': 'other'}

gearbox_corrections = {'manuell': 'manually', 'automatik': 'automatic'}

fuel_type_corrections = {'lpg': 'lpg', 'benzin': 'petrol',
                         'diesel': 'diesel', 'cng': 'cng',
                         'hybrid': 'hybrid', 'elektro':'electro',
                         'andere':'other'}

unrepaired_damage_corrections = {'nein':'no', 'ja': 'yes'}

# renaming values
autos['vehicle_type'] = autos['vehicle_type'].map(vehicle_type_corrections)

autos['gearbox'] = autos['gearbox'].map(gearbox_corrections)

autos['fuel_type'] = autos['fuel_type'].map(fuel_type_corrections)

autos['unrepaired_damage'] = autos['unrepaired_damage'].map(unrepaired_damage_corrections)

In [126]:
autos.head()

,date_crawled,name,price,abtest,vehicle_type,registration_year,gearbox,power_ps,model,odometer_km,registration_month,fuel_type,brand,unrepaired_damage,ad_created,postal_code,last_seen
0,2016-03-26 17:47:46,Peugeot_807_160_NAVTECH_ON_BOARD,5000,control,bus,2004,manually,158,andere,150000,3,lpg,peugeot,no,2016-03-26 00:00:00,79588,2016-04-06 06:45:54
1,2016-04-04 13:38:56,BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik,8500,control,limousine,1997,automatic,286,7er,150000,6,petrol,bmw,no,2016-04-04 00:00:00,71034,2016-04-06 14:45:08
2,2016-03-26 18:57:24,Volkswagen_Golf_1.6_United,8990,test,limousine,2009,manually,102,golf,70000,7,petrol,volkswagen,no,2016-03-26 00:00:00,35394,2016-04-06 20:15:37
3,2016-03-12 16:58:10,Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...,4350,control,small_car,2007,automatic,71,fortwo,70000,6,petrol,smart,no,2016-03-12 00:00:00,33729,2016-03-15 03:16:28
4,2016-04-01 14:38:50,Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfle...,1350,test,combi,2003,manually,0,focus,150000,7,petrol,ford,no,2016-04-01 00:00:00,39218,2016-04-01 14:38:50


We will convert the dates to be uniform numeric data, so "2016-03-26 17:47:46" becomes the integer "20160326".

In [127]:
# converting dates strings into integers
autos['date_crawled'] = autos['date_crawled'].str[:10].str.replace('-', '').astype(int)

autos['ad_created'] = autos['ad_created'].str[:10].str.replace('-', '').astype(int)

autos['last_seen'] = autos['last_seen'].str[:10].str.replace('-', '').astype(int)

# printing dates
print(autos[['date_crawled', 'ad_created', 'last_seen']])

       date_crawled  ad_created  last_seen
0          20160326    20160326   20160406
1          20160404    20160404   20160406
2          20160326    20160326   20160406
3          20160312    20160312   20160315
4          20160401    20160401   20160401
...             ...         ...        ...
49995      20160327    20160327   20160401
49996      20160328    20160328   20160402
49997      20160402    20160402   20160404
49998      20160308    20160308   20160405
49999      20160314    20160313   20160406

[34584 rows x 3 columns]


### 9.2. Analysis:

We will do:

- Find the most common `brand/model` combinations   
- Split the `odometer_km` into groups and use aggregation to see the average prices based on the mileage   
- Answer the question: How much cheaper are cars with damage than their non-damaged counterparts?

In [129]:
# finding the most common combinations
brand_model = autos.groupby(['brand','model'])['name'].count().sort_values(ascending=False).head(10)
brand_model    

brand          model   
volkswagen     golf        2749
bmw            3er         2083
volkswagen     polo        1133
opel           corsa       1090
               astra       1029
volkswagen     passat      1024
audi           a4           967
mercedes_benz  c_klasse     933
bmw            5er          922
mercedes_benz  e_klasse     764
Name: name, dtype: int64

According to the results, Volkswagen is the most popular brand with its models `golf`, `polo` and `passat` which sum `4,906` cars in total on the 10 most common brand/model.

The most common brand/model is `Volkswagen - Golf` with 2,749 cars, followed by `BMW - 3er` and `Volkswagen - Polo` with 2,083 and 1,133 respectively.

In [130]:
autos['odometer_km'].describe()

count     34584.000000
mean     124048.403886
std       39768.720586
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [131]:
# dataframe to analyze
mileage_price = autos[['odometer_km', 'price']]

# creating an dictionary to store our mean price
mieleage_intervals_km = {'0 - 30,000 km': 0, '30,000 - 60,000 km': 0, '60,000 - 90,000 km': 0, '90,000 - 120,000 km': 0, '120,000 - 150,000 km': 0}

# loop over the intervals
for interval in mieleage_intervals_km:    
    if interval == '0 - 30,000 km':
        km = mileage_price[mileage_price['odometer_km'] <= 30000]  
        mean_price = int(km['price'].mean())
        mieleage_intervals_km[interval] = mean_price  
        
    elif interval == '30,000 - 60,000 km':
        km = mileage_price[(mileage_price['odometer_km'] > 30000) & (mileage_price['odometer_km'] <= 60000)]  
        mean_price = int(km['price'].mean())
        mieleage_intervals_km[interval] = mean_price  
                            
    elif interval == '60,000 - 90,000 km':
        km = mileage_price[(mileage_price['odometer_km'] > 60000) & (mileage_price['odometer_km'] <= 90000)]  
        mean_price = int(km['price'].mean())
        mieleage_intervals_km[interval] = mean_price  
                            
    elif interval == '90,000 - 120,000 km':
        km = mileage_price[(mileage_price['odometer_km'] > 90000) & (mileage_price['odometer_km'] <= 120000)]  
        mean_price = int(km['price'].mean())
        mieleage_intervals_km[interval] = mean_price  
                            
    elif interval == '120,000 - 150,000 km':
        km = mileage_price[mileage_price['odometer_km'] > 120000]  
        mean_price = int(km['price'].mean())
        mieleage_intervals_km[interval] = mean_price
        
# converting dictionaries to series objects
mieleage_series = pd.Series(mieleage_intervals_km)

# creating a dataframe from the series objects
mieleage_data = pd.DataFrame(mieleage_series, columns=['mean_price'])
mieleage_data

,mean_price
"0 - 30,000 km",18494
"30,000 - 60,000 km",14224
"60,000 - 90,000 km",10232
"90,000 - 120,000 km",8927
"120,000 - 150,000 km",4637


As we can see in the dataframe, the mean price with respect to the mileage are indirectly proportional (The mean price decreases `$4,000` each `30,000 km`). 

In [132]:
# assigning the index of the brands
brands_index = autos['brand'].value_counts().index

# creating an empty dictionary to store our aggregate data
damage_no = {}
damage_yes = {}

# loop over the brands to calculate the mean price of non damage and damage cars
for each_brand in brands_index:
    brand_bool_no = autos[(autos['brand'] == each_brand) & (autos['unrepaired_damage'] == 'no')]
    brand_bool_yes = autos[(autos['brand'] == each_brand) & (autos['unrepaired_damage'] == 'yes')]
    
    mean_price_no = int(brand_bool_no['price'].mean())
    mean_price_yes = int(brand_bool_yes['price'].mean())
 
    damage_no[each_brand] = mean_price_no
    damage_yes[each_brand] = mean_price_yes

# converting dictionaries to series objects
damage_no_prices = pd.Series(damage_no)
damage_yes_prices = pd.Series(damage_yes)

# creating a dataframe from the series objects
brands_damage = pd.DataFrame(damage_no_prices, columns=['mean_price_non_damage'])

# assigning the series object in the dataframe
brands_damage['mean_price_damage'] = damage_yes_prices

# calculating the difference between prices
brands_damage['price_difference'] = brands_damage['mean_price_non_damage'] - brands_damage['mean_price_damage']

# printing the dataframe
brands_damage

,mean_price_non_damage,mean_price_damage,price_difference
volkswagen,6708,2366,4342
bmw,9722,3797,5925
mercedes_benz,10008,4282,5726
opel,3743,1464,2279
audi,11273,3587,7686
ford,4826,1548,3278
renault,3220,1200,2020
peugeot,3789,1463,2326
fiat,3565,1208,2357
seat,5338,1861,3477


In [133]:
# descriptive statistics
brands_damage['price_difference'].describe()

count       39.000000
mean      4647.051282
std       5896.032372
min        141.000000
25%       2309.500000
50%       3278.000000
75%       4226.500000
max      35787.000000
Name: price_difference, dtype: float64

As a result of this analysis, Every brand has the highest mean price in cars with non damage. Besides, the average price difference between non damage and damage cars is around `$4,647`.

In [134]:
autos

,date_crawled,name,price,abtest,vehicle_type,registration_year,gearbox,power_ps,model,odometer_km,registration_month,fuel_type,brand,unrepaired_damage,ad_created,postal_code,last_seen
0,20160326,Peugeot_807_160_NAVTECH_ON_BOARD,5000,control,bus,2004,manually,158,andere,150000,3,lpg,peugeot,no,20160326,79588,20160406
1,20160404,BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik,8500,control,limousine,1997,automatic,286,7er,150000,6,petrol,bmw,no,20160404,71034,20160406
2,20160326,Volkswagen_Golf_1.6_United,8990,test,limousine,2009,manually,102,golf,70000,7,petrol,volkswagen,no,20160326,35394,20160406
3,20160312,Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...,4350,control,small_car,2007,automatic,71,fortwo,70000,6,petrol,smart,no,20160312,33729,20160315
4,20160401,Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfle...,1350,test,combi,2003,manually,0,focus,150000,7,petrol,ford,no,20160401,39218,20160401
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
49995,20160327,Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon,24900,control,limousine,2011,automatic,239,q5,100000,1,diesel,audi,no,20160327,82131,20160401
49996,20160328,Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu...,1980,control,convertible,1996,manually,75,astra,150000,5,petrol,opel,no,20160328,44807,20160402
49997,20160402,Fiat_500_C_1.2_Dualogic_Lounge,13200,test,convertible,2014,automatic,69,500,5000,11,petrol,fiat,no,20160402,73430,20160404
49998,20160308,Audi_A3_2.0_TDI_Sportback_Ambition,22900,control,combi,2013,manually,150,a3,40000,11,diesel,audi,no,20160308,35683,20160405


## 10. Conclusions:

To sum up, the dataset of used cars from eBay Kleinanzeigenhe through this project was cleaned and analyzed using the tools and techniques learned during this course. First, we converted the column names from `camelcase` to `snakecase` due to Python prefers working with this practice of writing. In addition, the `price` and `odometer` columns were converted from string type to numeric type.

We identified German words in four columns in our dataset (`vehicle_type`, `gearbox`, `fuel_type` and `unrepaired_damage`) in order to translate it to English. Furthermore, these columns and `model` column contained null (or NaN) values that we dropped. Similarly, the columns `seller`, `nr_of_pictures` and `offer_type` were deleted from our dataset as well because they had same values which were not relevant for our analysis.

Besides these changes, we removed `1,421 cars` with the price of `$0` and `14 cars` which have values above `$350,000`, since the prices increased dramatically after this value. Regarding to when the cars were registered, we removed the values outside the `1900 - 2016 interval` because cars registered in the year 1000 or 9999 made no sense.

On the other hand, the keyword found out in the column `name` was the `cylinder capacity` which gave us the most common cylinder capacities (`2.0`, `1.6`, `1.4`).

In conclusion, we observed that in the top 20 brands `Audi`, `BMW` and `Mercedes Benz` are more expensive, `Renault` and `Fiat` are less expensive, and `Volkswagen` is in between. For this reason, `Volkswagen` is the most popular brand with its models `golf`, `polo` and `passat`. Additionally, the most common brand/model is `Volkswagen - Golf`, followed by `BMW - 3er` and `Volkswagen - Polo`. Although `audi`, `mercedes_benz` and `BMW` are the most expensives cars, these brands also have the highest mean mileage.

Furthermore, the price of each car in average decreases `$4,000` for every `30,000 km` driven and the average price difference between cars that are not damage and damage is around `$4,647`.
